rust/lance/src/dataset.rs

```diff
@@ -486,17 +488,63 @@ impl Dataset {
             });
         }
 
-        let fragments = write_fragments(
-            object_store.clone(),
-            &self.base,
-            &schema,
-            stream,
-            params.clone(),
-        )
-        .await?;
+        let transaction = match (params.tail_compaction_size, self.get_fragments().last()) {
+            (Some(tail_compact_thresh), Some(last_frag))
+                if last_frag.count_rows().await? <= tail_compact_thresh =>
+            {
+                let last_data = last_frag.scan().try_into_stream().await?;
+                let last_frag_stream: SendableRecordBatchStream = last_data.into();
+                let arrow_schema = last_frag_stream.schema();
+                let merged_stream = {
+                    // need to scope this use because TokioStreamExt collides with futures::StreamExt
+                    use tokio_stream::StreamExt as TokioStreamExt;
+                    RecordBatchStreamAdapter::new(arrow_schema, last_frag_stream.merge(stream))
```

Comment:
This merge does not preserve order.  I don't know if that is a problem or not but everything else we do respects the row order within fragments.